framework/auth/views.py

```diff
@@ -185,6 +191,33 @@ def auth_logout(redirect_url=None):
     return resp
 
 
+def auth_email_logout(token, user):
+    """When a user is adding an email or merging an account, add the email to the user and log them out.
+    """
+    redirect_url = web_url_for('auth_login') + '?existing_user={}'.format(urllib.quote_plus(user.email))
+    try:
+        unconfirmed_email = user.get_unconfirmed_email_for_token(token)
+    except KeyError:
+        raise HTTPError(http.BAD_REQUEST, data={
+            'message_short': "Bad token",
+            'message_long': "The provided token is invalid."
+        })
+    except ExpiredTokenError:
+        status.push_status_message('The private link you used is expired.')
+    try:
+        user_merge = User.find_one(Q('emails', 'eq', unconfirmed_email))
+    except NoResultsFound:
+        user_merge = False
+    if user_merge:
+        remove_sessions_for_user(user_merge)
+    user.email_verifications[token]['confirmed'] = True
```

Comment:
I tested this.  And you log in and get a message saying the user was not added, and revisit the link.  The merge won't happen until you go and click the correct link.